### PR TITLE
feat(web): degree progress chart

### PR DESCRIPTION
**Pull request by michaelzwang13**
**📌 What’s Changed** 

- added degree chart so that students can get a better look at the subject categorization of each degree
- added an overlay toggle so that students can look at how much of their degree they have completed

**✅ Actions**

- currently only shows the course code (CSCI-UA, EXPOS-UA, etc.) may want to consider writing (Computer Science, Expository Writing) for better clarity
- maybe add popup when hovering over the degree chart so users can look at the specific courses they need to take

**📝 Notes for Reviewer**

- i committed some mock data and seeding functions for my own testing purposes. i included it as it may be useful to the reviewers. once approved you or i can remove them
- frontend ui is not perfect, let me know if i can fix anything
- i can't seem to find my issue on the project board

Screenshot 2025-10-26 at 3 20 47 PM

### CODE REVIEW

**Comment by chenxin-yan** [comment]
🤖 **Manus AI Review**

## Summary

This PR introduces a degree progress chart feature to the web dashboard, allowing users to select academic programs and view their credit requirements and progress. It adds new components for program selection and chart visualization, integrates them into the dashboard page, and includes UI elements for toggling progress overlays. Additionally, it improves client-side rendering handling in the Convex client provider and adds reusable UI components for charts and checkboxes.

## Issues

🟡 **MEDIUM**: In ProgramRequirementsChart, the logic for matching user courses to program requirements assumes course codes are unique and directly comparable, but does not handle case sensitivity or partial matches, which could cause incorrect credit calculations.
   - *Location*: `apps/web/src/app/dashboard/components/charts/degree-charts.tsx`

🟡 **MEDIUM**: ProgramRequirementsChart does not handle the case where userCourses query returns null or errors explicitly; it only checks for undefined and truthiness, which might cause runtime errors or inconsistent UI states.
   - *Location*: `apps/web/src/app/dashboard/components/charts/degree-charts.tsx`

🟢 **LOW**: In ChartOverlayToggle component, the Checkbox input lacks an explicit id attribute, but the Label uses htmlFor='overlay', which may cause accessibility issues as the label is not properly linked to the input.
   - *Location*: `apps/web/src/app/dashboard/components/charts/label.tsx`

🟢 **LOW**: In ProgramSelector, the search input triggers a query on every keystroke without debouncing, which may cause excessive API calls and degrade performance.
   - *Location*: `apps/web/src/app/dashboard/components/charts/program-selector.tsx`

🟢 **LOW**: In ProgramSelector, the button elements for program selection use type='submit' inside a non-form context, which may cause unexpected form submission behavior if nested inside a form in the future.
   - *Location*: `apps/web/src/app/dashboard/components/charts/program-selector.tsx`

🟢 **LOW**: In ConvexClientProvider, the loading spinner uses hardcoded colors and styles that may not adapt well to dark mode or theming.
   - *Location*: `apps/web/src/components/ConvexClientProvider.tsx`

🟢 **LOW**: The new chart UI components in ui/chart.tsx are comprehensive but lack explicit TypeScript types for some internal helper functions, which could improve maintainability and type safety.
   - *Location*: `apps/web/src/components/ui/chart.tsx`

## Suggestions

💡 Normalize course codes (e.g., to uppercase) before comparison in ProgramRequirementsChart to avoid mismatches due to case differences.
   - *Location*: `apps/web/src/app/dashboard/components/charts/degree-charts.tsx`

💡 Add explicit error handling and loading states for userCourses query in ProgramRequirementsChart to improve robustness.
   - *Location*: `apps/web/src/app/dashboard/components/charts/degree-charts.tsx`

💡 Add an id attribute to the Checkbox in ChartOverlayToggle and update the Label's htmlFor accordingly to improve accessibility.
   - *Location*: `apps/web/src/app/dashboard/components/charts/label.tsx`

💡 Implement debouncing on the search input in ProgramSelector to reduce the number of API calls and improve performance.
   - *Location*: `apps/web/src/app/dashboard/components/charts/program-selector.tsx`

💡 Change the program selection buttons in ProgramSelector to type='button' to prevent unintended form submissions.
   - *Location*: `apps/web/src/app/dashboard/components/charts/program-selector.tsx`

💡 Consider using theme-aware colors or CSS variables for the loading spinner in ConvexClientProvider to support dark mode and theming.
   - *Location*: `apps/web/src/components/ConvexClientProvider.tsx`

💡 Add TypeScript types and JSDoc comments to internal helper functions in ui/chart.tsx for better maintainability.
   - *Location*: `apps/web/src/components/ui/chart.tsx`

💡 Add unit and integration tests covering edge cases such as empty programs, programs with zero credit requirements, and users with no completed courses.
   - *Location*: `apps/web/src/app/dashboard/components/charts/degree-charts.tsx`

💡 Add accessibility testing and keyboard navigation support for the ProgramSelector and ChartOverlayToggle components.
   - *Location*: `apps/web/src/app/dashboard/components/charts/program-selector.tsx and label.tsx`

## Positive Notes

✅ The PR introduces a well-structured and user-friendly UI for selecting programs and visualizing degree progress, enhancing the dashboard's functionality.

✅ The use of Recharts for data visualization is appropriate and leverages its features such as tooltips, labels, and stacked bars effectively.

✅ The code follows consistent React and TypeScript patterns, making use of hooks and functional components.

✅ The addition of a loading state and 'program not found' message improves the user experience during data fetching.

✅ The modular design with separate components for chart, label toggle, and program selector improves maintainability and reusability.

✅ The ConvexClientProvider update improves client-side rendering behavior by waiting for both mount and auth readiness before rendering children.

✅ The new UI components for checkbox and chart provide reusable building blocks consistent with the existing design system.

## Overall Assessment

This PR delivers a valuable feature with a clean and modular implementation that integrates well into the existing codebase. While the core logic and UI are sound, there are some medium and low severity issues related to input handling, accessibility, and performance optimizations that should be addressed to improve robustness and user experience. Adding more comprehensive testing and accessibility improvements would further enhance the quality. Overall, the PR is a solid addition that can be merged after addressing the noted concerns.

---

_This review was automatically generated by Manus AI. Please review and verify before committing to architectural decisions._

File: apps/web/src/app/dashboard/page.tsx [hunk 1]
```diff
@@ -1,7 +1,48 @@
...
+  if (isAuthenticated && student) {
+    for (const programId of student.programs) {
+      prgramQueries[programId] = {
+        query: api.programs.getProgramById,
+        args: { id: programId },
+      };
+    }
+  }
+
+  const programs = useQueries(prgramQueries);
```
Comment (chenxin-yan) [inline apps/web/src/app/dashboard/page.tsx]: we will be using the programs/majors that user selected during onboarding as the programs to display to users. I already set up all the queries for you.

File: packages/server/convex/programs.ts [hunk 2]
```diff
@@ -60,6 +60,84 @@ export const getProgramByName = protectedQuery({
   },
 });
 
+export const getProgramWithGroupedRequirements = protectedQuery({
```
Comment (chenxin-yan) [inline packages/server/convex/programs.ts]: considering deleting this server function and use getProgramByName (I already helped you set it up) and do the grouping/processing on the client side, so the data shape we got from the backend always reflect what is in our database.

File: apps/web/src/app/dashboard/page.tsx [hunk 3]
```diff
@@ -1,5 +1,73 @@
...
+    }
+  }
+
+  const programs: Record<
+    string,
+    FunctionReturnType<typeof api.programs.getProgramById>
+  > = useQueries(programQueries);
+
+  // Collect all unique course codes from all programs
+  const allCourseCodes = new Set<string>();
```
Comment (chenxin-yan) [inline apps/web/src/app/dashboard/page.tsx]: you are collecting all the course codes, but you should be treating courses with different types separately ("alternative", "required", "options")

File: apps/web/src/app/dashboard/components/degree-charts.tsx [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+  const program = Object.values(programs).find((p) => p !== undefined);
+  if (!program) return null;
+
+  // Calculate total credits for each category
+  const groupedRequirements: Record<
+    string,
+    { credits: number; courses: string[][] }
+  > = {};
+
+  // FIXME: currently no merging of programs if more than one listed (dual degree), assumes only one program for now
```
[the remark this answers is not in the document]
Comment (michaelzwang13) [reply apps/web/src/app/dashboard/components/degree-charts.tsx]: i'm not entirely sure what you mean by that. i currently already display the program name in the chart. are you saying that if a certain user has multiple programs (e.g. Computer Science B.A. and Economics B.A.), I should display it like "Program 1 and Program 2"? if that's the case i'll need to implement logic to merge the requirements of both programs. let me know what you think.

[the remark this answers is not in the document]
Comment (chenxin-yan) [reply apps/web/src/app/dashboard/components/degree-charts.tsx]: currently it only shows the program code instead of the actual program name: 
image

**Comment by chenxin-yan** [comment]
See #99 , `programName` is added to `courses` so you can map it to the chart

**Review by chenxin-yan — APPROVED** [review]
